### PR TITLE
use k8s atomic writer for storage of secrets / configmaps 

### PR DESCRIPTION
**Pull request by gabemontero**
Taking a glance at the latest shared resource epics, I cannot find a story related to this (@coreydaley @adambkaplan do let me know if that is just an oversight on my part), but an item we have discussed for completing productization of shared resource is switching actual storage of the secret/configmap keys to using [k8s atomic writer](https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/util/atomic_writer.go) to provide "atomic" access to data via some creative linux symlinking  so that if the contents change, users can use `inotify` or `fanotify` to coordinate access.

This is how a Pod's serviceaccount secrets have been stored since the early days of k8s, and this approach was created after some early bugs where intermittent errors would occur when read and writes of that secret data were done at the same time.

The upstream Secrets Store CSI Driver also adopted this approach somewhat recently.

As I mentioned last time we discussed this in team meetings, they actually create a copy of atomic_writer.go in their tree (even though they still already vendor in k8s/k8s).  In looking at the history of https://github.com/kubernetes-sigs/secrets-store-csi-driver/commits/main/pkg/util/fileutil/atomic_writer.go their updates from their initial copy are only superficial.

So I eschewed their copy approach and simply vendor in k8s/k8s like we do for openshift/builder and use the k8s/k8s atomic_writier.

Lastly, I moved our simple config map example to a locally created config map vs. one in the ocp config managed namespace.  It made it eaiser for manual testing around removing / adding keys and confirming that atomic_writer "handles it".

/assign @coreydaley 
/assign @adambkaplan

### CODE REVIEW

File: pkg/csidriver/driver.go [hunk 1]
```diff
@@ -22,18 +22,19 @@ import (
...
 	"strings"
 	"sync"
+	"syscall"
 
 	corev1 "k8s.io/api/core/v1"
 	kerrors "k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/klog/v2"
+	atomic "k8s.io/kubernetes/pkg/volume/util"
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:36]: I think that this should be something like "volumeutil", "kvolumeutil" or just "kutil" to be more descriptive to the package it is pulling in, not what feature of that package you are using.

File: pkg/csidriver/driver.go [hunk 2]
```diff
@@ -244,38 +245,43 @@ func commonUpsertRanger(dv *driverVolume, key, value interface{}) error {
...
 	// around host filesystem issues might already exist would be warranted with such an exploration/effort
 
-	// Next, on an update we first nuke any existing directory and then recreate it to simplify handling the case where
-	// the keys in the secret/configmap have changed such that some keys have been removed, which would translate
-	// in files having to be removed. commonOSRemove will handle shares mounted off of shares.  And a reminder,
-	// currently this driver does not support overlaying over directories with files.  Either the directory in the
-	// container image must be empty, or the directory does not exist, and is created for the Pod's container as
-	// part of provisioning the container.
-	if err := commonOSRemove(podPath, fmt.Sprintf("commonUpsertRanger key %s volid %s share id %s pod name %s", key, dv.GetVolID(), dv.GetSharedDataId(), dv.GetPodName())); err != nil {
+	// NOTE: atomic_writer handles any pruning of secret/configmap keys where were present before, but are no longer
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:248]: s/where/that/

File: pkg/csidriver/driver.go [hunk 3]
```diff
@@ -244,38 +245,43 @@ func commonUpsertRanger(dv *driverVolume, key, value interface{}) error {
...
+				Data: content,
+				Mode: 0644,
 			}
 		}
 	}
+	if len(podFile) > 0 {
+		err = aw.Write(podFile)
+		if err != nil {
+			return err
+		}
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:283]:
```
if err = aw.Write(podFile); err != nil {
	return err
}
```

File: pkg/csidriver/driver.go [hunk 4]
```diff
@@ -285,20 +291,48 @@ func commonOSRemove(dir, dbg string) error {
...
 	// off of the potential mount point
-	return filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
-		if info == nil {
-			return nil
-		}
-		// since we do not support mounting on existing content, a dir can only mean a share
-		// has been mounted as a separate dir in our share, so skip
-		if info.IsDir() {
-			return nil
+	dirFile, err := os.Open(dir)
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:294]: Should there be a deferred close directory after this?

File: pkg/csidriver/driver.go [hunk 5]
```diff
@@ -285,20 +291,48 @@ func commonOSRemove(dir, dbg string) error {
...
+		}
+		// then we delete any non symlinks
+		if dirEntry.Type()&os.ModeSymlink == 0 {
+			fileName := filepath.Join(dir, dirEntry.Name())
+			klog.V(4).Infof("commonOSRemove going to delete file %s", fileName)
+			err = os.RemoveAll(fileName)
+			if err != nil {
+				klog.V(0).Infof("commonOSRemove: %s", err.Error())
+				return err
+			}
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:332]:
```
if err = os.RemoveAll(fileName); err != nil {
klog.V(0).Infof("commonOSRemove: %s", err.Error())
return err
}
```
If this is going to be loglevel 0, it should probably be more human-friendly text, or maybe you meant it to be loglevel 4 like the others?

File: pkg/csidriver/driver_test.go [hunk 6]
```diff
@@ -825,14 +825,21 @@ func findSharedItems(t *testing.T, dir string) (bool, bool) {
...
 			return nil
 		}
-		t.Logf("found file %s dir flag %v", info.Name(), info.IsDir())
-		if err == nil && strings.Contains(info.Name(), secretkey1) {
+		searchName := info.Name()
+		if info.Mode()&os.ModeSymlink != 0 {
+			linkName, err := os.Readlink(info.Name())
+			if err == nil {
+				searchName = linkName
+			}
```
Comment (coreydaley) [inline pkg/csidriver/driver_test.go:833]:
```
if linkName, err := os.Readlink(info.Name()); err == nil {
    searchName = linkName
}
```

File: pkg/csidriver/driver.go [hunk 7]
```diff
@@ -244,38 +245,43 @@ func commonUpsertRanger(dv *driverVolume, key, value interface{}) error {
...
+	if err := os.MkdirAll(podPath, os.ModePerm); err != nil {
 		return err
 	}
-	if err := os.MkdirAll(podPath, os.ModePerm); err != nil {
+	podFile := map[string]atomic.FileProjection{}
+	aw, err := atomic.NewAtomicWriter(podPath, "shared-resource-csi-driver")
+	if err != nil {
 		return err
 	}
 	if payload.ByteData != nil {
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:258]: Can a `payload` have both `ByteData` and `StringData` populated at the same time? Can `payload` have both be `nil` also?

File: pkg/csidriver/driver.go [hunk 8]
```diff
@@ -285,20 +291,48 @@ func commonOSRemove(dir, dbg string) error {
...
+	}
+	// we have to unlink the key from our configmap/secret first with the atomic writer symlinks
+	for _, dirEntry := range dirEntries {
+		if dirEntry.Type()&os.ModeSymlink != 0 && !strings.HasPrefix(dirEntry.Name(), "..") {
+			// unlink our secret/configmap key
+			err = syscall.Unlink(filepath.Join(dir, dirEntry.Name()))
+			if err != nil {
+				klog.V(0).Infof("commonOSRemove: %s", err.Error())
+				return err
+			}
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:310]:
```
if err = syscall.Unlink(filepath.Join(dir, dirEntry.Name())); err != nil {
    klog.V(0).Infof("commonOSRemove: %s", err.Error())
    return err
}
```

This should probably be loglevel 4 or more human-friendly text.

File: pkg/csidriver/driver.go [hunk 9]
```diff
@@ -285,20 +291,48 @@ func commonOSRemove(dir, dbg string) error {
...
-	})
+	}
+	// we then unlink the symlink from atomic writer that start with ".."
+	for _, dirEntry := range dirEntries {
+		if dirEntry.Type()&os.ModeSymlink != 0 && strings.HasPrefix(dirEntry.Name(), "..") {
+			err = syscall.Unlink(filepath.Join(dir, dirEntry.Name()))
+			if err != nil {
+				klog.V(0).Infof("commonOSRemove: %s", err.Error())
+				return err
+			}
```
Comment (coreydaley) [inline pkg/csidriver/driver.go:321]:
```
if err = syscall.Unlink(filepath.Join(dir, dirEntry.Name())); err != nil {
    klog.V(0).Infof("commonOSRemove: %s", err.Error())
    return err
}
```
Loglevel 4 or more human-friendly text

**Review by coreydaley — CHANGES_REQUESTED** [review]

**Comment by gabemontero** [comment]
Compile errors I am not seeing locally:

```
Running e2e suite normal
# sigs.k8s.io/json/internal/golang/encoding/json
vendor/sigs.k8s.io/json/internal/golang/encoding/json/encode.go:1249:12: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
vendor/sigs.k8s.io/json/internal/golang/encoding/json/encode.go:1255:18: sf.IsExported undefined (type reflect.StructField has no field or method IsExported)
FAIL	github.com/openshift/csi-driver-shared-resource/test/e2e [build failed]
FAIL
```

probably need to bump the golang version in the Dockerfile and/or openshift/release

**Comment by gabemontero** [comment]
weirdly the unit test at least is using 1.16 which definitely has the issue:

[36mINFO[0m[2022-04-14T16:22:34Z] Tagging openshift/release:golang-1.16 into pipeline:root. 

e2e's are 1.17 ... though I cannot tell with dot rel off of 1.17

[36mINFO[0m[2022-04-14T16:22:35Z] Tagging ocp/builder:rhel-8-golang-1.17-openshift-4.11 into pipeline:ocp_builder_rhel-8-golang-1.17-openshift-4.11.

**Comment by gabemontero** [comment]
always forget the ci-operator yam file ... that at least fixed the unit test

**Comment by gabemontero** [comment]
removed permissions needs a little work ; my simple test did not catch what the more complicated e2e caught

will pick back up after the holiday

golang version at least sorted out

File: pkg/csidriver/driver.go (hunk 7 above)
Comment (gabemontero) [reply pkg/csidriver/driver.go:258]: yes and yes

see the ConfigMap godoc wrt both populated 

both nil means an empty secret/configmap

**Comment by gabemontero** [comment]
hmm ... slow passes for me locally

/test e2e-aws-csi-driver-slow

**Comment by gabemontero** [comment]
> hmm ... slow passes for me locally
> 
> /test e2e-aws-csi-driver-slow

reviewed ... the test waits 10 min and the relist is 10 min ... could have hit edge case.  May push tweak to wait time based on next set of results.

**Comment by gabemontero** [comment]
ok e2e's all green

doing one more run with the slow e2e interval tweaked

@coreydaley have the updates from your comments in a separate commit

PTAL / thanks

File: test/e2e/slow_test.go [hunk 10]
```diff
@@ -20,14 +20,14 @@ func TestBasicThenNoRBACThenRBAC(t *testing.T) {
...
+	t.Logf("%s: wait up to 11 minutes for examining pod %s since the controller does not currently watch all clusterroles and clusterrolebindings and reverse engineer which ones satisfied the SAR calls, so we wait for relist on shares", time.Now().String(), testArgs.Name)
 	testArgs.SearchStringMissing = true
-	testArgs.TestDuration = 10 * time.Minute
+	testArgs.TestDuration = 11 * time.Minute
 	testArgs.SearchString = "invoker"
 	framework.ExecPod(testArgs)
 
 	framework.CreateShareRelatedRBAC(testArgs)
-	t.Logf("%s: wait up to 10 minutes for examining pod %s since the controller does not currently watch all clusterroles and clusterrolebindings and reverse engineer which ones satisfied the SAR calls, so we wait for relist on shares", time.Now().String(), testArgs.Name)
+	t.Logf("%s: wait up to 11 minutes for examining pod %s since the controller does not currently watch all clusterroles and clusterrolebindings and reverse engineer which ones satisfied the SAR calls, so we wait for relist on shares", time.Now().String(), testArgs.Name)
```
Comment (coreydaley) [inline test/e2e/slow_test.go:30]: Does this need the `testArgs.TestDuration` or will it roll over from the previous one?

Comment (gabemontero) [reply test/e2e/slow_test.go:30]: rolls over from previous

Comment (coreydaley) [reply test/e2e/slow_test.go:30]: Ok, thanks.

**Comment by gabemontero** [comment]
> /lgtm

thanks @coreydaley 

that said,

/hold

have to squash your review commit

**Comment by gabemontero** [comment]
/label docs-approved

no docs impact - internal implementation detail - also see relative lack on docs in upstream k8s around using inotify, etc. ... considered a "linux" thing

**Comment by gabemontero** [comment]
/label px-approved

internal implementation detail

**Comment by gabemontero** [comment]
launching clusterbot cluster with this PR for final verification of entitled build usage

**Comment by gabemontero** [comment]
forgot, cannot do entitled builds with clusterbot clusters

had to verify manually via local build ... will do so again after this merges and we have a level with it

/label qe-approved

**Comment by gabemontero** [comment]
verified openshift builds with entitlements at level 4.11.0-0.ci-2022-04-18-211945
